velox/functions/prestosql/aggregates/benchmarks/ReduceAgg.cpp

```diff
@@ -0,0 +1,232 @@
+      results.push_back(result);
+    }
+
+    return copyResults(planFragment.planNode->outputType(), results);
+  }
+
+  RowVectorPtr copyResults(
+      const TypePtr& resultType,
+      const std::vector<VectorPtr>& results) {
+    if (results.empty()) {
+      return BaseVector::create<RowVector>(resultType, 0, pool());
+    }
+
+    auto totalCount = 0;
+    for (const auto& result : results) {
+      totalCount += result->size();
+    }
+
+    auto copy = BaseVector::create<RowVector>(resultType, totalCount, pool());
+    auto copyCount = 0;
+    for (const auto& result : results) {
+      copy->copy(result.get(), copyCount, 0, result->size());
+      copyCount += result->size();
+    }
+
+    return copy;
+  }
+
+  void runGlobal(const std::string& aggregate) {
+    folly::BenchmarkSuspender suspender;
+
+    auto plan = makeGlobalPlan(aggregate);
+    auto task = makeTask(plan);
+
+    suspender.dismiss();
+
+    vector_size_t numResultRows = 0;
+    while (auto result = task->next()) {
+      numResultRows += result->size();
+    }
+
+    folly::doNotOptimizeAway(numResultRows);
+  }
+
+  void runGroupBy(const std::string& aggregate) {
+    folly::BenchmarkSuspender suspender;
+
+    auto plan = makeGroupByPlan(aggregate);
+    auto task = makeTask(plan);
+
+    suspender.dismiss();
+
+    vector_size_t numResultRows = 0;
+    while (auto result = task->next()) {
+      numResultRows += result->size();
+    }
+
+    folly::doNotOptimizeAway(numResultRows);
+  }
+
+  core::PlanFragment makeGroupByPlan(const std::string& aggregate) {
+    return PlanBuilder()
+        .tableScan(inputType_)
+        .project({"i16 % 1234 as key", "i16"})
+        .partialAggregation({"key"}, {aggregate})
+        .finalAggregation()
+        .planFragment();
+  }
+
+  core::PlanFragment makeGlobalPlan(const std::string& aggregate) {
+    return PlanBuilder()
+        .tableScan(inputType_)
+        .partialAggregation({}, {aggregate})
+        .finalAggregation()
+        .planFragment();
+  }
+
+  std::shared_ptr<exec::Task> makeTask(core::PlanFragment plan) {
+    auto task = exec::Task::create(
+        "t",
+        std::move(plan),
+        0,
+        std::make_shared<core::QueryCtx>(executor_.get()));
+
+    task->addSplit("0", exec::Split(makeHiveConnectorSplit(filePath_->path)));
+    task->noMoreSplits("0");
+    return task;
+  }
+
+  RowTypePtr inputType_;
+  std::shared_ptr<TempFilePath> filePath_;
+};
+
+std::unique_ptr<ReduceAggBenchmark> benchmark;
+
+BENCHMARK(reduce_agg_global) {
+  benchmark->runGlobalReduceAgg();
+}
+
+BENCHMARK_RELATIVE(sum_global) {
+  benchmark->runGlobalSum();
+}
+
+BENCHMARK(reduce_agg_groupby) {
+  benchmark->runGroupByReduceAgg();
+}
+
+BENCHMARK_RELATIVE(sum_groupby) {
+  benchmark->runGroupBySum();
+}
+
+} // namespace
+
+int main(int argc, char** argv) {
+  folly::Init init(&argc, &argv);
+
+  benchmark = std::make_unique<ReduceAggBenchmark>();
+  benchmark->verify();
```

Comment:
This step is to verify benchmark is correct before we actually run the registered benchmarks?

Reply:
Yes. This is to ensure that query plan we compare produce same results.